src/typechecker.erl

```diff
@@ -3438,16 +3438,71 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+
+-spec check_guard_call(atom(), list()) -> map().
+check_guard_call(is_atom, [{var, _, Var}]) -> #{Var => type(atom)};
+check_guard_call(is_binary, [{var, _, Var}]) -> #{Var => type(binary)};
+check_guard_call(is_bitstring, [{var, _, Var}]) -> #{Var => type(bitstring)};
+check_guard_call(is_boolean, [{var, _, Var}]) -> #{Var => type(boolean)};
+check_guard_call(is_float, [{var, _, Var}]) -> #{Var => type(float)};
+check_guard_call(is_function, [{var, _, Var}]) -> #{Var => type('fun')};
+check_guard_call(is_function, [{var, _, Var}, {integer, _, Arity}]) -> #{Var => type_fun(Arity)};
+check_guard_call(is_integer, [{var, _, Var}]) -> #{Var => type(integer)};
+check_guard_call(is_list, [{var, _, Var}]) -> #{Var => type(list)};
+check_guard_call(is_map, [{var, _, Var}]) -> #{Var => type(map)};
+check_guard_call(is_number, [{var, _, Var}]) -> #{Var => type(number)};
+check_guard_call(is_pid, [{var, _, Var}]) -> #{Var => type(pid)};
+check_guard_call(is_port, [{var, _, Var}]) -> #{Var => type(port)};
+check_guard_call(is_record, [{var, _, Var}, {atom, _, Record}]) -> #{Var => type_record(Record)};
+check_guard_call(is_record, [{var, _, Var}, {atom, _, Record}, _]) -> #{Var => type_record(Record)};
+check_guard_call(is_reference, [{var, _, Var}]) -> #{Var => type(reference)};
+check_guard_call(is_tuple, [{var, _, Var}]) -> #{Var => type(tuple)};
+check_guard_call(_Fun, _Vars) -> #{}.
+
+-spec check_guard(#env{}, term()) -> map().
+check_guard(_Env, {call, _, {atom, _, Fun}, Vars}) ->
+    check_guard_call(Fun, Vars);
+check_guard(_Env, {call, _, {remote,_,_,{atom, _, Fun}}, Vars}) ->
+    check_guard_call(Fun, Vars);
+check_guard(Env, {op, _OrElseAnno, 'orelse', Call1, Call2}) ->
+    G1 = check_guard(Env, Call1),
+    G2 = check_guard(Env, Call2),
+    union_var_binds_symmetrical([G1, G2], Env#env.tenv);
+check_guard(Env, {op, _OrAnno, 'or', Call1, Call2}) ->
+    G1 = check_guard(Env, Call1),
+    G2 = check_guard(Env, Call2),
+    union_var_binds_symmetrical([G1, G2], Env#env.tenv);
```

Comment:
This duplicated code can easily be combined into

```Erlang
check_guard(Env, {op, _Anno, Op, Call1, Call2}) when Op == 'orelse';
                                                     Op == 'or' ->
    G1 = check_guard(Env, Call1),
    G2 = check_guard(Env, Call2),
    union_var_binds_symmetrical([G1, G2], Env#env.tenv);
```

and similarily for 'and' and 'andalso'.

Reply:
Done